pkg/udf/rpc/grpc_batch_map.go

```diff
@@ -0,0 +1,217 @@
+
+	// Read routine: this goroutine iterates over the input messages and sends each
+	// of the read messages to the grpc client after transforming it to a BatchMapRequest.
+	// Once all messages are sent, it closes the input channel to indicate that all requests have been read.
+	// On creating a new request, we add it to a tracker map so that the responses on the stream
+	// can be mapped backed to the given parent request
+	go func() {
+		defer close(inputChan)
+		for _, msg := range messages {
+			trackerReq.addRequest(msg)
+			inputChan <- u.parseInputRequest(msg)
+		}
+	}()
+
+	// Process the responses received on the response channel:
+	// This is an infinite loop which would exit
+	// 1. Once there are no more responses left to read from the channel
+	// 2. There is an error received on the error channel
+	// We have not added a case for context.Done as there is a handler for that in the client, and it should
+	// propagate it on the error channel itself.
+	//
+	// On getting a response, it would parse them and create a new ReadWriteMessagePair entry in the udfResults
+	// Any errors received from the client, are propagated back to the caller.
+loop:
+	for {
+		select {
+		// got an error on the error channel, so return immediately
+		case err := <-errCh:
+			err = &ApplyUDFErr{
+				UserUDFErr: false,
+				Message:    fmt.Sprintf("gRPC client.BatchMapFn failed, %s", err),
+				InternalErr: InternalErr{
+					Flag:        true,
+					MainCarDown: false,
+				},
+			}
+			return nil, err
+		case grpcResp, ok := <-respCh:
+			// if there are no more messages to read on the channel we can break
+			if !ok {
+				break loop
+			}
+			// Get the unique request ID for which these responses are meant for.
+			msgId := grpcResp.GetId()
+			// Fetch the request value for the given ID from the tracker
+			parentMessage, ok := trackerReq.getRequest(msgId)
+			if !ok {
+				// this case is when the given request ID was not present in the tracker.
+				// This means that either the UDF added an incorrect ID
+				// This cannot be processed further and should result in an error
+				// Can there be another case for this?
+				logger.Error("Request missing from tracker, ", msgId)
+				return nil, fmt.Errorf("incorrect ID found during batch map processing")
+			}
+			// parse the responses received
+			// TODO(map-batch): should we make this concurrent by using multiple goroutines, instead of sequential.
+			// Try and see if any perf improvements from this.
+			parsedResp := u.parseResponse(grpcResp, parentMessage)
+			responsePair := isb.ReadWriteMessagePair{
+				ReadMessage:   parentMessage,
+				WriteMessages: parsedResp,
+				Err:           nil,
+			}
+			udfResults = append(udfResults, responsePair)
+			trackerReq.removeRequest(msgId)
+		}
+	}
+	// check if there are elements left in the tracker. This cannot be an acceptable case as we want the
+	// UDF to send responses for all elements.
+	leftRequest := trackerReq.getItems()
```

Comment:
It looks like  a `isEmpty()` function is good enough, do not need to have a `getItems()`, right?

Reply:
Sure! I can update that.